src/tool/cppwinrt/test/tearoff.cpp

```diff
@@ -0,0 +1,277 @@
+#include "pch.h"
+#include <windows.foundation.h>
+
+namespace abi
+{
+    using namespace ABI::Windows::Foundation;
+}
+
+namespace winrt
+{
+    using namespace Windows::Foundation;
+}
+
+namespace
+{
+    // Creates an implementation of IPersist as a tearoff.
+    HRESULT make_persist(winrt::IUnknown const& object, void** result) noexcept
+    {
+        struct persist final : IPersist
+        {
+            persist(winrt::IUnknown const& object) :
+                m_object(object.as<::IUnknown>())
+            {
+            }
+
+            HRESULT __stdcall GetClassID(CLSID* result) noexcept final
+            {
+                *result = winrt::guid_of<IPersist>();
+                return S_OK;
+            }
+
+            HRESULT __stdcall QueryInterface(GUID const& id, void** result) noexcept final
+            {
+                if (winrt::is_guid_of<IPersist>(id))
+                {
+                    *result = static_cast<IPersist*>(this);
+                    AddRef();
+                    return S_OK;
+                }
+
+                return m_object->QueryInterface(id, result);
+            }
+
+            ULONG __stdcall AddRef() noexcept final
+            {
+                return 1 + m_references.fetch_add(1, std::memory_order_relaxed);
+            }
+
+            ULONG __stdcall Release() noexcept final
+            {
+                uint32_t const remaining = m_references.fetch_sub(1, std::memory_order_relaxed) - 1;
+
+                if (remaining == 0)
+                {
+                    delete this;
+                }
+
+                return remaining;
+            }
+
+        private:
+
+            winrt::com_ptr<::IUnknown> m_object;
+            std::atomic<uint32_t> m_references{ 1 };
+        };
+
+        *result = new (std::nothrow) persist(object);
+        return *result ? S_OK : E_OUTOFMEMORY;
+    }
+
+    // Creates an implementation of IStringable as a tearoff.
+    HRESULT make_stringable(winrt::IInspectable const& object, winrt::hstring const& value, void** result) noexcept
+    {
+        struct stringable final : abi::IStringable
```

Comment:
could factor a base_tearoff for m_object mgmt, but if these are the only intended test cases, don't bother

Reply:
Yes, just examples really.